Source/Diagnostics/ReducedDiags/ColliderRelevant.cpp

```diff
@@ -0,0 +1,572 @@
+        add_diag("x_std_"+species_names[i_s], "x_std_"+species_names[i_s]+"(m)");
+        add_diag("thetax_min_"+species_names[i_s], "theta_x_min_"+species_names[i_s]+"(rad)");
+        add_diag("thetax_ave_"+species_names[i_s], "theta_x_ave_"+species_names[i_s]+"(rad)");
+        add_diag("thetax_max_"+species_names[i_s], "theta_x_max_"+species_names[i_s]+"(rad)");
+        add_diag("thetax_std_"+species_names[i_s], "theta_x_std_"+species_names[i_s]+"(rad)");
+#endif
+        m_data.resize(all_diag_names.size());
+    }
+
+    if (amrex::ParallelDescriptor::IOProcessor())
+    {
+        if ( m_write_header )
+        {
+            // open file
+            std::ofstream ofs;
+            ofs.open(m_path + m_rd_name + "." + m_extension,
+                std::ofstream::out | std::ofstream::app);
+            // write header row
+            int off = 0;
+            ofs << "#";
+            ofs << "[" << off++ << "]step()";
+            ofs << m_sep;
+            ofs << "[" << off++ << "]time(s)";
+            for (const auto& name : all_diag_names){
+                const auto& el = m_headers_indices[name];
+                ofs << m_sep << "[" << el.idx + off << "]" << el.header;
+            }
+            ofs << std::endl;
+            // close file
+            ofs.close();
+        }
+    }
+}
+
+void ColliderRelevant::ComputeDiags (int step)
+{
+#if defined(WARPX_DIM_RZ)
+    amrex::ignore_unused(step);
+#else
+    // Judge if the diags should be done
+    if (!m_intervals.contains(step+1)) { return; }
+
+    // get MultiParticleContainer class object
+    const MultiParticleContainer& mypc = WarpX::GetInstance().GetPartContainer();
+
+    // get species names (std::vector<std::string>)
+    const std::vector<std::string> species_names = mypc.GetSpeciesNames();
+
+    // get a reference to WarpX instance
+    auto& warpx = WarpX::GetInstance();
+
+    // get cell size
+    amrex::Geometry const & geom = warpx.Geom(0);
+#if defined(WARPX_DIM_1D_Z)
+        amrex::Real dV = geom.CellSize(0);
+#elif defined(WARPX_DIM_XZ)
+        amrex::Real dV = geom.CellSize(0) * geom.CellSize(1);
+#elif defined(WARPX_DIM_3D)
+        amrex::Real dV = geom.CellSize(0) * geom.CellSize(1) * geom.CellSize(2);
+#endif
+
+    const auto get_idx = [&](const std::string& name){
+        return m_headers_indices.at(name).idx;
+    };
+
+    std::array<std::unique_ptr<amrex::MultiFab>, 2> num_dens;
+
+    // loop over species
+    for (int i_s = 0; i_s < 2; ++i_s)
+    {
+        // get WarpXParticleContainer class object
+        WarpXParticleContainer& myspc = mypc.GetParticleContainerFromName(m_beam_name[i_s]);
+        // get charge
+        amrex::ParticleReal const q = myspc.getCharge();
+
+        using PType = typename WarpXParticleContainer::SuperParticleType;
+
+        num_dens[i_s] = myspc.GetChargeDensity(0);
+        num_dens[i_s]->mult(1./q);
+
+#if defined(WARPX_DIM_XZ)
```

Comment:
I'm not sure if it would be better, but could these separate blocks of code be rewritten intertwined since there is a fair amount of code duplication with completely separate blocks for 2D and 3D?

Reply:
@EZoni has been looking into this, and it seems that the code would become quite fragmented with a lot of `#if defined` statements.

Reply:
Use of `AMREX_D_TERM` and `AMREX_D_DECL` might help.

Reply:
Hi Dave, I had tried considering those to reduce duplication, but somehow it still did not seem that feasible/readable overall.

The reason is that there are parts of the code where we have, for instance,
```cpp
auto r_std = amrex::ParticleReduce<amrex::ReduceData<Real, Real>>
```
in 2D, and
```cpp
auto r_std = amrex::ParticleReduce<amrex::ReduceData<Real, Real, Real, Real>>(
```
in 3D.

You see that here the reduce syntax needs 2 args in 2D and 4 args in 3D: I don't know how we could handle this in a dimension-agnostic way using `AMREX_D_TERM` or `AMREX_D_DECL`, which map only 1D to 1 arg, 2D to 2 args, 3D to 3 args. Do you see an easy solution for such cases?

Reply:
Hence I thought that we could investigate such a refactoring, which didn't seem trivial to me, possibly later in a separate PR, to avoid stalling this one for too long.

Reply:
Doing the refactoring in a subsequent PR is fine.

I think you can do things like `AMREX_D_DECL(,Real,Real)`, so you can do this
```amrex::ParticleReduce<amrex::ReduceData<AMREX_D_DECL(,Real,Real),AMREX_D_DECL(,Real,Real)>```
I do agree that it starts to get messy.